upload_multipart.go

```diff
@@ -0,0 +1,238 @@
+	return cli.UploadPartWithPsk(ctx, req, payload, nil)
+}
+
+// UploadPartWithPsk handles the uploading a file to AWS S3, into the bucket configured for this client, using a user-defined psk
+func (cli *Client) UploadPartWithPsk(ctx context.Context, req *UploadPartRequest, payload []byte, psk []byte) error {
+	logData := log.Data{
+		"chunk_number": req.ChunkNumber,
+		"max_chunks":   req.TotalChunks,
+		"file_name":    req.FileName,
+		"bucket_name":  cli.bucketName,
+		"user_psk":     psk != nil,
+	}
+
+	// Get UploadID or create it if it does not exist (atomically)
+	uploadID, err := cli.doGetOrCreateMultipartUpload(ctx, req)
+	if err != nil {
+		return NewError(err, logData)
+	}
+
+	// Do the upload against AWS
+	_, err = cli.doUploadPart(
+		ctx,
+		&s3.UploadPartInput{
+			UploadId:   &uploadID,
+			Bucket:     &cli.bucketName,
+			Key:        &req.UploadKey,
+			Body:       bytes.NewReader(payload),
+			PartNumber: &req.ChunkNumber,
+		},
+		psk,
+	)
+	if err != nil {
+		return NewError(err, logData)
+	}
+
+	log.Info(ctx, "chunk accepted", logData)
+
+	// List parts so that we can validate if the upload operation is complete
+	output, err := cli.sdkClient.ListParts(
+		&s3.ListPartsInput{
+			Key:      &req.UploadKey,
+			Bucket:   &cli.bucketName,
+			UploadId: &uploadID,
+		},
+	)
+	if err != nil {
+		return NewError(fmt.Errorf("error listing parts: %w", err), logData)
+	}
+
+	// If all parts have been uploaded, we call completeUpload
+	parts := output.Parts
+	if len(parts) == req.TotalChunks {
+		return cli.completeUpload(ctx, uploadID, req, parts)
+	}
+
+	// Otherwise we don't need to perform any other operation.
+	return nil
+}
+
+// doGetOrCreateMultipartUpload atomically gets the UploadId for the specified bucket
+// and S3 object key, and if it does not find it, it creates it.
+func (cli *Client) doGetOrCreateMultipartUpload(ctx context.Context, req *UploadPartRequest) (uploadID string, err error) {
+	cli.mutexUploadID.Lock()
+	defer cli.mutexUploadID.Unlock()
+
+	// List existing Multipart uploads for our bucket
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(
+		&s3.ListMultipartUploadsInput{
+			Bucket: &cli.bucketName,
+		})
+	if err != nil {
+		return "", fmt.Errorf("error fetching multipart list: %w", err)
+	}
+
+	// Try to find a multipart upload for the same s3 object that we want
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+			return uploadID, nil
+		}
+	}
+
+	// If we didn't find the Multipart upload, create it
+	createMultiOutput, err := cli.sdkClient.CreateMultipartUpload(
+		&s3.CreateMultipartUploadInput{
+			Bucket:      &cli.bucketName,
+			Key:         &req.UploadKey,
+			ContentType: &req.Type,
+		})
+	if err != nil {
+		return "", fmt.Errorf("error creating multipart upload: %w", err)
+	}
+	return *createMultiOutput.UploadId, nil
```

Comment:
The code logic here looks absolutely fine.

However, I've noticed this method & several others in this file are declaring _named return vars_ in the func signature but you are not always returning those values - For example:

```
if *upload.Key == req.UploadKey {
	uploadID = *upload.UploadId
	return uploadID, nil
}
```
returns 1 of the named vars and nil. Whereas

```
return *createMultiOutput.UploadId, nil
```
Is returning neither of the named vars.

I am not a fan of the  _named return vars_  func style I don't think it is always clear what the value of each var will be when you `return`. I think if we are going to use that style it's important to be consistent and ensure we are always returning the named return variables e.g.
```
if *upload.Key == req.UploadKey {
	uploadID = *upload.UploadId
	return
} 
```
Otherwise I think its potentially misleading and harder to fully understand - the func signature is promising one thing but we're not necessarily meeting that promise.

I'm aware this is me being picky but I think consistency and clarity are really important for long term maintainability so I think it is worth addressing. Happy to discuss.

Reply:
I will remove the var names from the signature, I think it will be clearer. As this is a private function, I think it will not cause confusion.